src/connectors/voice-search/connectVoiceSearch.ts

```diff
@@ -0,0 +1,112 @@
+import voiceSearchHelper from '../../lib/voiceSearchHelper';
+import { VoiceListeningState } from '../../widgets/voice-search/voice-search';
+
+const withUsage = createDocumentationMessageGenerator({
+  name: 'voice-search',
+  connector: true,
+});
+
+export interface VoiceSearchRenderOptions<T> extends RenderOptions<T> {
+  isSupportedBrowser: boolean;
+  isListening: boolean;
+  toggleListening: (toggleListeningParams: {
+    searchAsYouSpeak: boolean;
+  }) => void;
+  voiceListeningState: VoiceListeningState;
+}
+
+export type VoiceSearchRenderer<T> = Renderer<VoiceSearchRenderOptions<T>>;
+
+export type VoiceSearchWidgetFactory<T> = WidgetFactory<T>;
+
+export type VoiceSearchConnector = <T>(
+  renderFn: VoiceSearchRenderer<T>,
+  unmountFn?: () => void
+) => VoiceSearchWidgetFactory<T>;
+
+const connectVoiceSearch: VoiceSearchConnector = (
+  renderFn,
+  unmountFn = noop
+) => {
+  checkRendering(renderFn, withUsage());
+
+  return (widgetParams = {}) => {
+    const render = ({
+      isFirstRendering,
+      instantSearchInstance,
+      voiceSearchHelper: {
+        isSupportedBrowser,
+        isListening,
+        toggleListening,
+        getState,
+      },
+    }) => {
+      renderFn(
+        {
+          isSupportedBrowser: isSupportedBrowser(),
+          isListening: isListening(),
+          toggleListening,
+          voiceListeningState: getState(),
+          widgetParams,
+          instantSearchInstance,
+        },
+        isFirstRendering
+      );
+    };
+
+    return {
+      init({ helper, instantSearchInstance }) {
+        (this as any)._refine = (() => {
+          let previousQuery: string;
+          const setQueryAndSearch = (query: string) => {
+            if (query !== helper.state.query) {
+              previousQuery = helper.state.query;
+              helper.setQuery(query);
+            }
+            if (
+              typeof previousQuery !== 'undefined' &&
+              previousQuery !== query
+            ) {
+              helper.search();
+            }
+          };
+          return setQueryAndSearch;
+        })();
+        (this as any)._voiceSearchHelper = voiceSearchHelper({
+          onQueryChange: query => (this as any)._refine(query),
+          onStateChange: () => {
+            render({
+              isFirstRendering: false,
+              instantSearchInstance,
+              voiceSearchHelper: (this as any)._voiceSearchHelper,
+            });
+          },
+        });
+        render({
+          isFirstRendering: true,
+          instantSearchInstance,
+          voiceSearchHelper: (this as any)._voiceSearchHelper,
+        });
+      },
+      render({ instantSearchInstance }) {
+        render({
+          isFirstRendering: false,
+          instantSearchInstance,
+          voiceSearchHelper: (this as any)._voiceSearchHelper,
+        });
+      },
+      dispose({ state }) {
+        unmountFn();
+        return state.setQuery('');
+      },
+    };
```

Comment:
I didn't know. After reading [this](https://www.algolia.com/doc/guides/building-search-ui/widgets/create-your-own-widgets/js/?language=react#the-widget-lifecycle-and-api), I guess I need to define `getWidgetState` and `getWidgetSearchParameters`, right?
And in case of VoiceSearch, the implementations will be the same as SearchBox.
Let me know if I'm wrong about this. @samouss cc @Haroenv

Reply:
You're right. We can postpone it to a latter PR though because this one is already consequent.

Reply:
Done